plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiClassBuilderInterceptor.kt

```diff
@@ -200,9 +239,31 @@ class JvmAbiClassBuilderInterceptor(
     }
 }
 
-private fun shouldRemoveFromAbi(irClass: IrClass?, removePrivateClasses: Boolean): Boolean = when {
+private fun shouldRemoveFromAbi(irClass: IrClass?, removePrivateClasses: Boolean, removeInternalDeclarations: Boolean): Boolean = when {
     irClass == null -> false
     irClass.isFileClass -> false
-    removePrivateClasses -> irClass.isEffectivelyPrivate()
+    removePrivateClasses -> irClass.isVisibilityStrippedFromAbi(stripInternal = removeInternalDeclarations)
     else -> false
 }
+
+private fun IrDeclarationWithVisibility.isVisibilityStrippedFromAbi(stripInternal: Boolean): Boolean {
+    val isInAbi = visibility == DescriptorVisibilities.PUBLIC
+            || visibility == DescriptorVisibilities.PROTECTED
+            || (!stripInternal && visibility == DescriptorVisibilities.INTERNAL)
+
+    return when {
+        isInAbi -> parentClassOrNull?.isVisibilityStrippedFromAbi(stripInternal) ?: false
+
+        visibility == DescriptorVisibilities.INVISIBLE_FAKE -> {
```

Comment:
o_O. How is it possible to obtain invisible_fake here? I'm not sure what this code is even doing...